tests/performance/density.go

```diff
@@ -0,0 +1,316 @@
+		By(fmt.Sprintf("Creating VMI %s", vmi.ObjectMeta.Name))
+		_, err := virtClient.VirtualMachineInstance(util.NamespaceTestDefault).Create(vmi)
+		Expect(err).ToNot(HaveOccurred())
+
+		// control the throughput with waiting interval
+		interval := testArg.ArrivalRateFunc(testArg.ArrivalRate)
+		time.Sleep(interval * testArg.ArrivalUnit)
+	}
+}
+
+func watchVMI(virtClient kubecli.KubevirtClient, status *e2emetrics.BatchStatus, testArg DensityTest, quit chan bool) {
+	defer GinkgoRecover()
+
+	watcher, err := virtClient.VirtualMachineInstance(util.NamespaceTestDefault).Watch(metav1.ListOptions{})
+	Expect(err).ToNot(HaveOccurred())
+
+	// wait the VMI deletion status
+	timeout := time.After(testArg.VMBatchDeletionLimit * time.Second)
+
+	for {
+		select {
+		case <-quit:
+			return
+		case <-timeout:
+			err = fmt.Errorf("failed to create %v VMIs within the expected interval %v",
+				testArg.NumVMs, testArg.VMBatchDeletionLimit*time.Second)
+			Expect(err).ToNot(HaveOccurred())
+
+		case event := <-watcher.ResultChan():
+			handleVMIEvent(event, status, virtClient, testArg)
+		}
+	}
+}
+
+func handleVMIEvent(event watch.Event, status *e2emetrics.BatchStatus, virtClient kubecli.KubevirtClient, testArg DensityTest) error {
+	defer GinkgoRecover()
+
+	vmi, ok := event.Object.(*kvv1.VirtualMachineInstance)
+	Expect(ok).To(BeTrue())
+
+	switch event.Type {
+
+	case watch.Modified:
+		if vmi.Status.Phase == kvv1.Running {
+			status.AddVMIPhaseMetrics(*vmi)
+		}
+
+	case watch.Deleted:
+		go waitVMIDisapear(virtClient, vmi, status, testArg.VMDeletionLimit*time.Second)
+	}
+
+	return nil
+}
+
+// waitVMIDisapear measure the deletion time without watching events due to scalability reasons
+func waitVMIDisapear(virtClient kubecli.KubevirtClient, vmi *kvv1.VirtualMachineInstance, status *e2emetrics.BatchStatus, timeout time.Duration) {
+	defer GinkgoRecover()
+
+	// for high accuracy, we watch the phases with a very small pooling interval
+	pollingInterval := 300 * time.Millisecond
+
+	By(fmt.Sprintf("Waiting VMI %s disapear", vmi.ObjectMeta.Name))
+	Eventually(func() error {
+		_, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.ObjectMeta.Name, &metav1.GetOptions{})
+		return err
+	}, timeout, pollingInterval).Should(
+		SatisfyAll(HaveOccurred(), WithTransform(errors.IsNotFound, BeTrue())),
+		"The VMI should be gone within the given timeout",
+	)
+
+	status.AddVMIDeletionMetrics(*vmi)
+}
+
+func deleteVMIBatch(virtClient kubecli.KubevirtClient, status *e2emetrics.BatchStatus, testArg DensityTest) {
+	status.RLock()
+	for _, data := range status.GetVMIPhaseMap(kvv1.Running) {
+		// the GracePeriod is set to force the server to set the DeletionTimestamp
+		gracePeriodSeconds := int64(0)
+		err := virtClient.VirtualMachineInstance(data.Namespace).Delete(data.Name, &metav1.DeleteOptions{GracePeriodSeconds: &gracePeriodSeconds})
+		Expect(err).To(BeNil())
+		time.Sleep(200 * time.Millisecond)
```

Comment:
What is this sleep for?

Reply:
The sleep here is to avoid to overload the Kubernetes API server with too much requests and become the bottleneck of the analysis.  Additionally, it also prevent the k8s client go to apply rate limiting (throttling) in the requests.

Reply:
Makes sense, thanks for the explanation. Maybe we could add comment here: `// Sleep to manually throttle API requests.`